File: Python/pytree/pytree_lib/scan.py
"""Directory scanning and size formatting."""
from __future__ import annotations

import os
from pathlib import Path
from typing import Optional

from pytree_lib.models import DirInfo, DirSummary, ProgressCb, ScanStats, format_size

__all__ = ["format_size", "scan_directory", "get_dir_size"]


def scan_directory(
    path: Path,
    max_depth: Optional[int] = None,
    current_depth: int = 0,
    *,
    stats: Optional[ScanStats] = None,
    progress_cb: Optional[ProgressCb] = None,
) -> DirInfo:
    if stats is None:
        stats = ScanStats()
    total_size = 0
    file_count = 0
    dir_count = 0
    children: list[DirInfo] = []
    error = None

    if progress_cb is not None:
        stats.current = str(path)
        progress_cb(stats)

    try:
        with os.scandir(path) as it:
            entries = list(it)
    except PermissionError:
        error = "Permission denied"
        entries = []
    except OSError as e:
        error = str(e)
        entries = []

    for entry in entries:
        if entry.is_symlink():
            continue
        item = Path(entry.path)
        try:
            if entry.is_file(follow_symlinks=False):
                sz = entry.stat(follow_symlinks=False).st_size
                total_size += sz
                file_count += 1
                stats.files += 1
                stats.size += sz
                if progress_cb is not None:
                    progress_cb(stats)
                children.append(
                    DirInfo(
                        path=item,
                        size=sz,
                        file_count=1,
                        dir_count=0,
                        children=[],
                    )
                )
            elif entry.is_dir(follow_symlinks=False):
                dir_count += 1
                stats.dirs += 1
                if max_depth is None or current_depth < max_depth:
                    child_info = scan_directory(
                        item,
                        max_depth,
                        current_depth + 1,
                        stats=stats,
                        progress_cb=progress_cb,
                    )
                    children.append(child_info)
                    total_size += child_info.size
                    file_count += child_info.file_count
                    dir_count += child_info.dir_count
                else:
                    summary = get_dir_size(item, stats=stats, progress_cb=progress_cb)
                    child_info = DirInfo(
                        path=item,
                        size=summary.size,
                        file_count=summary.files,
                        dir_count=summary.dirs,
                        children=[],
                        error=summary.error,
                    )
                    children.append(child_info)
                    total_size += summary.size
                    file_count += summary.files
                    dir_count += summary.dirs
        except PermissionError:
            continue
        except OSError:
            continue

    children.sort(key=lambda x: x.size, reverse=True)
    return DirInfo(
        path=path,
        size=total_size,
        file_count=file_count,
        dir_count=dir_count,
        children=children,
        error=error,
    )


def get_dir_size(
    path: Path,
    *,
    stats: Optional[ScanStats] = None,
    progress_cb: Optional[ProgressCb] = None,
) -> DirSummary:
    summary = DirSummary()
    try:
        for root, dirnames, filenames in os.walk(path, followlinks=False):
            root_path = Path(root)
            for name in filenames:
                item = root_path / name
                try:
                    if item.is_symlink():
                        continue
                    sz = item.stat().st_size
                    summary.size += sz
                    summary.files += 1
                    if stats is not None:
                        stats.files += 1
                        stats.size += sz
                        if progress_cb is not None:
                            progress_cb(stats)
                except OSError:
                    continue
            summary.dirs += len(dirnames)
            if stats is not None:
                stats.dirs += len(dirnames)
    except OSError as e:
        summary.error = f"{type(e).__name__}: {e}"
    return summary

File: Rust/rust-sizetree/src/scan.rs
use std::fs;
use std::path::Path;

use anyhow::Result;

use crate::models::{DirInfo, DirSummary, ScanStats};
use crate::progress::ScanProgress;

#[derive(Clone, Copy, Debug, Default)]
pub struct ScanOptions {
    pub max_depth: Option<u32>,
    pub include_hidden: bool,
}

pub fn scan_directory(
    path: &Path,
    max_depth: Option<u32>,
    current_depth: u32,
) -> Result<DirInfo> {
    scan_directory_with(
        path,
        &ScanOptions {
            max_depth,
            include_hidden: false,
        },
        current_depth,
        None,
    )
}

pub fn scan_directory_with(
    path: &Path,
    opts: &ScanOptions,
    current_depth: u32,
    progress: Option<&mut ScanProgress>,
) -> Result<DirInfo> {
    let mut progress = progress;
    let mut stats = ScanStats::default();
    Ok(scan_directory_inner(
        path,
        opts,
        current_depth,
        &mut stats,
        &mut progress,
    ))
}

fn scan_directory_inner(
    path: &Path,
    opts: &ScanOptions,
    current_depth: u32,
    stats: &mut ScanStats,
    progress: &mut Option<&mut ScanProgress>,
) -> DirInfo {
    stats.current = Some(path.display().to_string());
    notify_progress(progress, stats);

    let mut total_size = 0u64;
    let mut file_count = 0u64;
    let mut dir_count = 0u64;
    let mut children = Vec::new();
    let mut error = None;

    let entries = match fs::read_dir(path) {
        Ok(it) => it.filter_map(Result::ok).collect::<Vec<_>>(),
        Err(e) if e.kind() == std::io::ErrorKind::PermissionDenied => {
            error = Some("Permission denied".to_string());
            Vec::new()
        }
        Err(e) => {
            error = Some(e.to_string());
            Vec::new()
        }
    };

    for entry in entries {
        let name = entry.file_name();
        let name_str = name.to_string_lossy();
        if !opts.include_hidden && entry_is_hidden(&entry, &name_str) {
            continue;
        }

        let item = entry.path();
        let ft = match entry.file_type() {
            Ok(ft) => ft,
            Err(_) => continue,
        };
        if ft.is_symlink() {
            continue;
        }

        if ft.is_file() {
            match entry.metadata() {
                Ok(meta) => {
                    let sz = meta.len();
                    total_size += sz;
                    file_count += 1;
                    stats.files += 1;
                    stats.size += sz;
                    notify_progress(progress, stats);
                    children.push(DirInfo {
                        path: item,
                        size: sz,
                        file_count: 1,
                        dir_count: 0,
                        children: vec![],
                        error: None,
                    });
                }
                Err(_) => continue,
            }
        } else if ft.is_dir() {
            dir_count += 1;
            stats.dirs += 1;
            let child_info = if opts.max_depth.is_none() || current_depth < opts.max_depth.unwrap_or(0)
            {
                scan_directory_inner(&item, opts, current_depth + 1, stats, progress)
            } else {
                let summary = get_dir_size(&item, stats, progress);
                DirInfo {
                    path: item,
                    size: summary.size,
                    file_count: summary.files,
                    dir_count: summary.dirs,
                    children: vec![],
                    error: summary.error,
                }
            };
            total_size += child_info.size;
            file_count += child_info.file_count;
            dir_count += child_info.dir_count;
            children.push(child_info);
        }
    }

    children.sort_by(|a, b| b.size.cmp(&a.size));
    DirInfo {
        path: path.to_path_buf(),
        size: total_size,
        file_count,
        dir_count,
        children,
        error,
    }
}

fn entry_is_hidden(entry: &fs::DirEntry, name: &str) -> bool {
    if name.starts_with('.') {
        return true;
    }
    #[cfg(windows)]
    {
        use std::os::windows::fs::MetadataExt;
        if let Ok(meta) = entry.metadata() {
            const FILE_ATTRIBUTE_HIDDEN: u32 = 0x2;
            if meta.file_attributes() & FILE_ATTRIBUTE_HIDDEN != 0 {
                return true;
            }
        }
    }
    false
}

pub fn get_dir_size(
    path: &Path,
    stats: &mut ScanStats,
    progress: &mut Option<&mut ScanProgress>,
) -> DirSummary {
    let mut summary = DirSummary {
        size: 0,
        files: 0,
        dirs: 0,
        error: None,
    };
    walk_dir_size(path, &mut summary, stats, progress);
    summary
}

fn walk_dir_size(
    path: &Path,
    summary: &mut DirSummary,
    stats: &mut ScanStats,
    progress: &mut Option<&mut ScanProgress>,
) {
    stats.current = Some(path.display().to_string());
    notify_progress(progress, stats);

    let entries = match fs::read_dir(path) {
        Ok(it) => it,
        Err(e) => {
            summary.error = Some(e.to_string());
            return;
        }
    };
    for entry in entries.flatten() {
        let ft = match entry.file_type() {
            Ok(ft) => ft,
            Err(_) => continue,
        };
        if ft.is_symlink() {
            continue;
        }
        if ft.is_dir() {
            summary.dirs += 1;
            stats.dirs += 1;
            walk_dir_size(&entry.path(), summary, stats, progress);
        } else if ft.is_file() {
            if let Ok(meta) = entry.metadata() {
                let sz = meta.len();
                summary.size += sz;
                summary.files += 1;
                stats.files += 1;
                stats.size += sz;
                notify_progress(progress, stats);
            }
        }
    }
}

fn notify_progress(progress: &mut Option<&mut ScanProgress>, stats: &ScanStats) {
    if let Some(p) = progress.as_deref_mut() {
        p.notify(stats);
    }
}
